oidc_middleware.go

```diff
@@ -0,0 +1,87 @@
+package main
+
+import (
+	"crypto/rsa"
+	"encoding/base64"
+	"encoding/json"
+	"github.com/gin-gonic/gin"
+	"math/big"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/golang-jwt/jwt"
+)
+// Largely taken from https://developer.okta.com/blog/2021/01/04/offline-jwt-validation-with-go
+
+func GetOidcPublicKeys(config OICDConfig) *map[string]*rsa.PublicKey {
+	rsakeys := make(map[string]*rsa.PublicKey)
+	var body map[string]interface{}
+	resp, err := http.Get(config.KeysURL)
+	if err != nil {
+		log.Error().WithError(err).Message("Could not fetch from KeysURL.")
+	}
+	err = json.NewDecoder(resp.Body).Decode(&body)
+	if err != nil {
+		log.Error().WithError(err).Message("Failed to decode login JWT Keys.")
+	}
```

Comment:
Strongly suggest using a static type instead of `map[string]interface{}`

Reply:
Where and for what? And what type should i use instead?

Reply:
A custom made one. By looking at the implementation, my guess is that something like this would work:

```go
type OIDCKeyResponse struct {
    Keys []struct{
        KeyID  string `json:"kid"`
        Number string `json:"n"`
    } `json:"keys"`
}

var body OIDCKeyResponse
err = json.NewDecoder(resp.Body).Decode(&body)
```

Reply:
> Defining a strict type here seems like it could cause type errors on unexpect input where we should not be handling the errors in the first place.

Don't know if there's a miscommunication here with what I'm proposing.

This variable that you're defining a few lines above:

https://github.com/iver-wharf/wharf-api/blob/4546dd92368b1dba73b05c288367bc50a55235db/oidc_middleware.go#L39

My point is to make it type-safe with the fields you're already assuming to be of certain names and types. The lot of `.([]interface{})` and `.(string)` casting you're doing can fail in panics.

Instead I'm proposing to use a type-safe static type with the fields you're unmarshalling, as suggested in the code sample in my comment earlier in this thread.

> I think it would be best left up to the jwt handler library to define and work with the type returned for the keys, its a part of what its there for.

This does not regard any JWT handler library. You're only using the `encoding/json` library.

Reply:
Resolved in pairprogramming session